### PR TITLE
[TwoAddressInstruction] Propagate undef flags for partial defs

### PR DESCRIPTION
**Pull request by perlfu**
If part of a register (lowered from REG_SEQUENCE) is undefined then we should propagate undef flags to uses of those lanes. This is only performed when live intervals are present as it requires live intervals to correctly match uses to defs, and the primary goal is to allow precise computation of subrange intervals.

### CODE REVIEW

**Review by jayfoad — COMMENTED** [review]
I guess this is required for correctness, but it is not clear to me what the failure is without this. Do you have a small MIR example that needs this?

**Comment by perlfu** [comment]
I've switch this to a concrete MIR example.

The root cause could be seen as an issue/bug with GlobalISel code gen, although I think this change is still beneficial.
In this case there is an extractelement on a v7 type, which must be widened to v8 for code gen:
`%ext = extractelement <7 x double> %vec, i32 %sel`

The result is that GlobalISel generates a `REG_SEQUENCE` with an undef (see last element):
```
%34:vreg_512 = REG_SEQUENCE %16, %subreg.sub0_sub1, %17, %subreg.sub2_sub3, %18, %subreg.sub4_sub5, %19, %subreg.sub6_sub7, %20, %subreg.sub8_sub9, %21, %subreg.sub10_sub11, %22, %subreg.sub12_sub13, undef %35:vreg_64, %subreg.sub14_sub15
```

Then generates code to extract that element if selected:
```   
%80:vgpr_32 = V_CNDMASK_B32_e64 0, %76, 0, %34.sub14, %79, implicit $exec
%81:vgpr_32 = V_CNDMASK_B32_e64 0, %77, 0, %34.sub15, %79, implicit $exec
```

Without propagating the undef flags, the above instructions will fail the verifier if LiveIntervals are available.

File: llvm/test/CodeGen/AMDGPU/GlobalISel/twoaddr-extract-dyn-v7f64.mir [hunk 1]
```diff
@@ -0,0 +1,134 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 4
+# RUN: llc -mtriple=amdgcn-mesa-mesa3d -mcpu=gfx900 -early-live-intervals -run-pass=liveintervals -run-pass=twoaddressinstruction -verify-machineinstrs -o - | FileCheck %s
```
Comment (jayfoad) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/twoaddr-extract-dyn-v7f64.mir:2]:
```suggestion
# RUN: llc -mtriple=amdgcn-mesa-mesa3d -mcpu=gfx900 -early-live-intervals -run-pass=liveintervals -run-pass=twoaddressinstruction -verify-machineinstrs -o - %s | FileCheck %s
```

File: llvm/lib/CodeGen/TwoAddressInstructionPass.cpp [hunk 2]
```diff
@@ -1991,11 +1997,25 @@ eliminateRegSequence(MachineBasicBlock::iterator &MBBI) {
       MI.removeOperand(j);
   } else {
     if (LIS) {
-      // Force interval recomputation if we moved from full definition
-      // of register to partial.
-      if (DefIsPartial && LIS->hasInterval(DstReg) &&
-          MRI->shouldTrackSubRegLiveness(DstReg))
+      // Force live interval recomputation if we moved to a partial defintion
```
Comment (arsenm) [inline llvm/lib/CodeGen/TwoAddressInstructionPass.cpp:2000]:
```suggestion
      // Force live interval recomputation if we moved to a partial definition
```

File: llvm/test/CodeGen/AMDGPU/GlobalISel/twoaddr-extract-dyn-v7f64.mir [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+--- |
+  define double @dyn_extract_v7f64_v_v(<7 x double> %vec, i32 %sel) #0 {
+  entry:
+    %ext = extractelement <7 x double> %vec, i32 %sel
+    ret double %ext
+  }
+
+  attributes #0 = { "amdgpu-no-completion-action" "amdgpu-no-default-queue" "amdgpu-no-dispatch-id" "amdgpu-no-dispatch-ptr" "amdgpu-no-heap-ptr" "amdgpu-no-hostcall-ptr" "amdgpu-no-implicitarg-ptr" "amdgpu-no-lds-kernel-id" "amdgpu-no-multigrid-sync-arg" "amdgpu-no-queue-ptr" "amdgpu-no-workgroup-id-x" "amdgpu-no-workgroup-id-y" "amdgpu-no-workgroup-id-z" "amdgpu-no-workitem-id-x" "amdgpu-no-workitem-id-y" "amdgpu-no-workitem-id-z" "amdgpu-waves-per-eu"="4,10" "target-cpu"="gfx900" "uniform-work-group-size"="false" }
+...
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/GlobalISel/twoaddr-extract-dyn-v7f64.mir:12]: Don't think you need the IR section

**Comment by perlfu** [comment]
- Rebase
- Address reviewer comments: remove test IR, fix comment typo

**Review by arsenm — APPROVED** [review]

**Pull request merged**